[DNM] add actual keys read to command queue

tbg opened the pull request:
The code is a mess and not meant to go in; don't bother leaving style
comments. Look only at the second commit.

This is an exploration for #6290. Briefly put, evaluation of commands
before proposing their result to Raft exercises the command queue in ways
the current code did not (which executes everything serially because it's
on the Raft goroutine).

In particular, the command queue now has to live up to its role of allowing
only commands which can be parallelized to be evaluated (and subsequently
proposed and applied).

What the command queue tracked so far are simply the main key ranges of the
requests in the batch, but that is not nearly enough.

To err on the side of being conservative, this PR presents an (unmergeably
hacky) attempt to classify what keys really need to be protected, as
measured by our testing coverage. As such, this list is not complete and in
particular it does not take into account reads of Replica state where the
in-memory cache is hit (though that could be caught by not accessing state
directly but having it go to disk every time; not done yet).

Major takeaways are:
- Splits need to lock the key range of the RHS descriptor for writes (we
  can
  get a little tricky with reads), but also touches all of the abort cache
  for
  reading. It also accesses the Replica state for the RHS, though that's
  a lesser concern.
- Every command needs to read FrozenStatus, RangeLastGCKey and writes
  MVCCStats. Unless we are smart about it, that already completely kills off
  any proposal parallelism. We will have to be smart about it because that's
  pretty bad. Especially the stats update is interesting because that is
  both
  frequent and a write operation. The obvious solution here is to make use
  of
  the commutativity of stats delta updates (though `ContainsEstimates` needs
  to
  be taken into account conservatively).
- Every transaction reads its AbortCacheKey and reads or writes its
  transaction
  record (see discussion in #6290 on why the latter makes sense).

The proposal's MaxLeaseIndex already makes sure that when we submit command
A to Raft before command B, then they can only apply in that order (mod
refurbishments, which can reorder and cause problems to this arguments,
though I think they can be overcome). So we could imagine a world in which
the command queue wouldn't commonly block commands on top of each other:
- a command arrives.
- checks the command queue; finds that it must wait for some prior commands
  to
  apply.
- instead of waiting, it evaluates these commands in the correct order and
  applies itself on top of that, entering itself in the command queue but
  immediately proposing to Raft a command whose successful application is
  contingent upon the previous commands applying as planned.
- once the command applies, leave the command queue (as usual)
- if the command does not apply (i.e. it or its dependencies got
  reordered),
  leave the command queue and enter it anew (instead of the previous
  internal
  refurbishment cycle).

The above is fairly half baked and challenging to implement, but so is
tracking every command's implicit writes; the above idea would in principle
remove any latency that isn't Raft or doing actual work, which is a great
benefit.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10084)

<!-- Reviewable:end -->

petermattis commented:
> Every command needs to read FrozenStatus, RangeLastGCKey and writes MVCCStats.

For the MVCCStats writes, I was assuming that was why the proposal was containing a delta. 

Why does every write op have to read `RangeLastGCKey`?

petermattis commented:
> instead of waiting, it evaluates these commands in the correct order and applies itself on top of that, entering itself in the command queue but immediately proposing to Raft a command whose successful application is contingent upon the previous commands applying as planned.

This is tricky because in order to evaluate a command on top of a previous command we would need to have the previous command's batch (or we would have to reevaluate the previous command which feels like a non-starter). I suppose I can imagine a hierarchy of batches. This would require a bit of worker in `storage/engine` to accomplish. Perhaps I'm misunderstanding what you are thinking of here.

tschottdorf commented:
> This is tricky because in order to evaluate a command on top of a previous command we would need to have the previous command's batch (or we would have to reevaluate the previous command which feels like a non-starter). I suppose I can imagine a hierarchy of batches. This would require a bit of worker in storage/engine to accomplish. Perhaps I'm misunderstanding what you are thinking of here.

The simpler one would be the former, but of course the reevaluation cost seems prohibitive. The tree of `engine.Batch` is what I pictured, though that does sound kind of spooky.

> Why does every write op have to read RangeLastGCKey?

Sorry, not `RangeLastGCKey`, but `RangeGCThresholdKey` (and the requirement could likely be relaxed, but definitely a `CPut` would have to).

petermattis commented:
> Sorry, not RangeLastGCKey, but RangeGCThresholdKey

`RangeTxnSpanGCThresholdKey`? Seems like that value only ratchets forward. 

> The above is fairly half baked and challenging to implement, but so is
tracking every command's implicit writes.

Can we move the implicit reads and writes into `ProposalData`? What is the scope here? Is it too onerous/fragile to identify all of them?

tschottdorf commented:
> RangeTxnSpanGCThresholdKey? Seems like that value only ratchets forward.

Yes, but you can't issue a `CPut` at a timestamp that is lower than the threshold at the time you propose that `CPut` but higher when that `CPut` applies.

tschottdorf commented:
> Can we move the implicit reads and writes into ProposalData? What is the scope here? Is it too onerous/fragile to identify all of them?

What do you mean by moving them?
Yeah, it's pretty onerous and fragile. The evaluation gets a `Batch` and gets to do whatever it pleases, plus there's access to the in-memory state that should also count towards accessing keys.

The need to special-case and be clever about a lot of this stuff because the trivial solution is wildly unperformant in the common case is pretty frightening.

petermattis commented:
> What do you mean by moving them?

Any key which is being touched (read or written) on every command would not be added to the `WriteBatch` but would instead have some flag and/or special handling in `ProposalData`. Onerous to find all these locations, but we could add a wrapper around `engine.Batch` which prohibits writing these keys.

I'm probably speculating a bit too much here without having thought through the problem thoroughly. I expect massive holes to be poked in the above.

tschottdorf commented:
Yeah, that amounts to moving more logic downstream of Raft again, and all of these problems have a solution, but I'd rather they have a solution with a solid abstraction so that it's manageable. Regardless, even the batch-tree needs to know what is going to be touched before evaluation happens, so it doesn't really change the game in that regard. I wish there were a better solution than looking hard at each command before it goes into the command queue, but it can probably be made workable.

tschottdorf commented:
@nvanbenschoten going in the direction this PR is headed, we would add a bunch of highly contended keys to the command queue which however are most often only read. Also, there are always range-local keys in there - I believe we do some shady things in the command queue that may now backfire by creating unnecessarily large spans that always conflict with each other:

```
	// Create the covering entry. Note that this may have an "illegal" key range
	// spanning from range-local to range-global, but that's acceptable here as
	// long as we're careful in the future.
```

Do you have input on how to deal with both of those effects? Would having a range-local component with tighter semantics to the command queue make sense? The rough idea would be that state access would act like a `RWMutex`, i.e. in normal operation everyone just grabs read locks and nothing really happens. A split would grab the exclusive lock, which of course blocks everyone after, so splits would be more expensive. Or we could "honestly" track the reads and writes of state of each command, but then we need to revisit the optimizations.

petermattis commented on pkg/storage/replica.go:
```diff
@@ -1143,15 +1144,104 @@ func (r *Replica) checkBatchRequest(ba roachpb.BatchRequest) error {
+		// the heuristic we are using in the CommandQueue to make a large
+		// span first is going to be horrible as we are always including
+		// range-local and range-global keys, and so that mutant span includes
+		// the whole world.
```
Probably need to pass in separate range-local and range-global spans to the `CommandQueue`.

tschottdorf replied:
Agreed. We'll want to be much more fine grained about what's read-only and what's read-write (much of the system spans are almost always accessed read-only).

petermattis commented on pkg/storage/replica.go:
```diff
@@ -1143,15 +1144,104 @@ func (r *Replica) checkBatchRequest(ba roachpb.BatchRequest) error {
+				spans = append(spans, roachpb.Span{
+					Key: keys.RangeLastGCKey(r.RangeID),
+				}, roachpb.Span{
+					Key: keys.RangeLastGCKey(r.RangeID),
```
I assume the duplication of `RangeLastGCKey` was an accident.

tschottdorf replied:
Yep.

petermattis commented:
> Yeah, that amounts to moving more logic downstream of Raft again, and all of these problems have a solution, but I'd rather they have a solution with a solid abstraction so that it's manageable.  

"For every complex problem there is an answer that is clear, simple, and wrong." H. L. Mencken

Good point about moving logic downstream of Raft again. I'm still thinking about this.

bdarnell commented on pkg/storage/replica.go:
```diff
@@ -1143,15 +1144,104 @@ func (r *Replica) checkBatchRequest(ba roachpb.BatchRequest) error {
+			spans = append(spans, roachpb.Span{
+				Key: keys.AbortCacheKey(r.RangeID, ba.Txn.ID),
+			}, roachpb.Span{
+				Key: keys.TransactionKey(ba.Txn.Key, ba.Txn.ID),
```
I understand now why we need to read the abort cache key, but we shouldn't be reading the transaction record on every command - it might not be present on this range.

tschottdorf replied:
You're right, will address this.

bdarnell reviewed (commented):
I don't see the stats key or the GC threshold key for every command. Are those just left out of this PR?

tschottdorf commented:
re @bdarnell in [#6290](https://github.com/cockroachdb/cockroach/issues/6290#issuecomment-255013778):

> We have two different ordering mechanisms: the command queue and the lease index. Since the lease index gives a total order to all commands, maybe we don't need a tree of batches but a linear chain: every command is proposed on top of the batch produced by the previous one (and whenever we reset our lease index we rewind to whatever is on disk and start over). Would this free us from having to figure out which commands could interfere with each other for the purposes of proposer-evaluated KV? (this wouldn't let us get rid of the command queue, of course, but it would mean we wouldn't have to expand its role or add additional complications)

The list vs. tree thing is a tradeoff, but I think the list is too simplistic (?).

You would presumably queue all the writes back-to-back, i.e. there's a single list of batches. The downside here is that one write has to wait for a completely unrelated write to create its Batch, i.e. while a large DeleteRange is evaluated, a lot of writes to a different part of the Range has to just sit there.

Reads still have to know fairly fine-grained information from the command queue. For example, reads need to know if something is in-flight that mutates, say, the GC threshold. Naively, that means that an inflight `GCRequest` blocks all reads until it applies, which is, of course, unnecessary in all but contrived practical situations and a high tax for those. Ideally an inflight GCRequest blocks *only* the reads at a timestamp that violate the GCThreshold that this request would set, but this is a very fine-grained condition. Easy to do, but it would be done through a custom data structure. I think similar considerations for the other fields will still let us gravitate towards maintaining fine-grained control over which key ranges are accessed by which command.

Back to proposals: The list version above is fairly bad when an expensive operation blocks non-overlapping operations. For example, picture a huge `DeleteRange` which comes in before a lot of individual non-overlapping single key writes. In that situation, we would put `DeleteRange` in the command queue, and then start evaluating it. All the following writes would have to sit out that evaluation, which isn't necessary. They could look at the command queue and realize that they don't overlap, and could evaluate themselves in parallel (since they also all mutually don't overlap), and propose themselves in the order in which they get there. Perhaps at some point the `DeleteRange` would finish and propose itself; all legit, because the command queue guarantees that all of these proposals commute. That corresponds to the direction in this PR (assuming the system keys are suitably special-cased) and requires all the fine-grained tracking of key ranges.

Adding the batch hierarchy to that results in the tree of batches. The case in which this shines is when the `DeleteRange` has evaluated and now more writes are piling up on top of that. We save the consensus latency that way, which is of course huge. But we absolutely need to keep tracking the "real" set of keys touched by commands for it to be correct.

But now that I've considered all of these options above, it seems that

1. the list of batches obviates perhaps a little of the tracking complexity (but not all) but is probably bad for the high percentile latencies (or are we afraid of too much parallelism at proposal time to hurt us? We could of course limit the number of concurrent evaluations)
2. the command tracking approach here might be the most beneficial to set up now: We won't be able to avoid it completely, and it precisely avoids the pitfalls of 1.
3. the tree of batches is pretty advanced and there seems no good reason to pursue this now.

My vote at this point is to make this PR work, that is to suitably special-case those keys which are touched by almost every command (`FrozenStatus`, `{TxnSpan,}GCThreshold`). Ideally at some point we would not give a raw `Batch` to the commands but some tighter interface (think of a `Batch` that has a more high-level interface on what keys are being touched, and which can prevent access to keys it wasn't set up for), but for the inital version it would be based on trust and assertions as presented in this PR.

I'll go and annotate all of the necessary spans found in the PR with possible strategies and their remaining performance impact. Then I'll spend some time looking at the mostly orthogonal raftMu question again to give folks some time to take a look at this post and the update to the PR (will ping when that has happened).

@petermattis Creating a hierarchy of `Batch` wouldn't be more difficult than

```go
b1 := e.NewBatch()
doStuff(b1)
b2 := e.NewBatch()
b2.ApplyBatchRepr(b1.Repr())
doStuff(b2)
```

With some sugar added to populate `b2` directly from `b1` mod the round-trip through Go? Maintaining the right "base" to create the batches from sounds like an interesting problem, though.

petermattis commented:
> ```
b2 := e.NewBatch()
b2.ApplyBatchRepr(b1.Repr())
doStuff(b2)
```

At this point, `b2.Repr()` contains all of the `b1` operations as well as the `b2` operations. Don't you want it to only contain the `b2` operations? Since batches are append-only we could probably do some hackery to extract out the `b2` mutations.

Also note that `b2.ApplyBatchRepr(b1.Repr())` is moderately expensive as it has to re-index all of the keys in `b1`. Certainly much less expensive than re-performing the operation for `b1`, but definitely not free. 

I was imaging we'd do something like:

```
b1 := e.NewBatch()
doStuff(b1)
b2 := b1.NewBatch()
doStuff(b2)
```

This wouldn't work at the moment because you can't create a batch on top of another batch. In particular, iterators in such a list of batches would need to be fixed (see `db.cc:DBBatch::NewIter`). This approach would also make iterators progressively more expensive as the depth of the list increased.

tschottdorf commented:
> I was imaging we'd do something like:

Yes, that's what it would look like in code (though really the Batch should be unusable once it's branched off of, so perhaps that API isn't great), but I was trying to understand what would go on under the hood. Doesn't seem worth poking this too much now as I don't expect us to actually try to do batch compounding.

tschottdorf commented:
Re @bdarnell:

> I don't see the stats key or the GC threshold key for every command. Are those just left out of this PR?

`Distinct()` writes weren't tracked. They are now. We'll need a better API to make it harder for Raft commands to access random parts of keyspace.

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1214 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10084#-KUYhiqlKYxIDikyrx1z:-KUYhiqlKYxIDikyrx2-:b-y4sinl) ([raw file](https://github.com/cockroachdb/cockroach/blob/725e79fa6b17b39cb26ec8bfdf81c6a8bc51243c/pkg/storage/replica.go#L1214)):*
> ```Go
> 				})
> 			case *roachpb.HeartbeatTxnRequest:
> 				if ba.Txn == nil {
> ```

When does this occur?

---

*[pkg/storage/replica.go, line 1235 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10084#-KUYh_bj0W582V-RCxQf:-KUYh_bj0W582V-RCxQg:b-s7vly5) ([raw file](https://github.com/cockroachdb/cockroach/blob/725e79fa6b17b39cb26ec8bfdf81c6a8bc51243c/pkg/storage/replica.go#L1235)):*
> ```Go
> 				}, roachpb.Span{
> 					// This one isn't asserted but serves as a reminder that
> 					// something odd might be going on when we the commit
> ```

Some grammar is off here: `when we the commit`.

---

*[pkg/storage/replica.go, line 1303 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10084#-KUYi9S8LZ6Y-geUkGT5:-KUYi9S8LZ6Y-geUkGT6:b3iluk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/725e79fa6b17b39cb26ec8bfdf81c6a8bc51243c/pkg/storage/replica.go#L1303)):*
> ```Go
> 				// This blocks all other commands, but that's fine.
> 				spans = append(spans, roachpb.Span{ // rw
> 					Key: keys.RangeFrozenStatusKey(r.RangeID),
> ```

You already added this above. Before making this change, I think you should do another PR which changes `CommandQueue.{getWait,add}` to take separate slices of read and write spans. And there is another question as to whether the API should separate out range-local vs global keys or if `CommandQueue` should make that determination internally. I'd lean towards the latter. 

---

*[pkg/storage/replica.go, line 1373 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10084#-KUYigGzmORACdN4NwFq:-KUYigGzmORACdN4NwFr:b-j1rwjo) ([raw file](https://github.com/cockroachdb/cockroach/blob/725e79fa6b17b39cb26ec8bfdf81c6a8bc51243c/pkg/storage/replica.go#L1373)):*
> ```Go
> 	return func(br *roachpb.BatchResponse, pErr *roachpb.Error, debug interface{}) *roachpb.Error {
> 		// TODO(tschottdorf): check that our spans added above cover everything
> 		// we touched. In particilar, each individual write must be covered
> ```

`s/particilar/particular/g`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10084)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Just talked to @tschottdorf offline about this and one of the major takeaways was that splitting up the command queue for range-local and range-global keys seems to be the way to go. This will allow the standard range-global command queue to behave as it does now without having to reconsider previous optimizations.

Neither of us are confident that an interval tree is the best structure for the new range-local cache. The workload will be almost entirely read-only, which means that the overhead of searching an interval tree for all overlaps, only to throw most of them away, would be incredibly inefficient. On top of this, most (all?) requests will be working on a single key, not a key range. An initial idea is to have a map from keys to lists of pending kv operations for that key, with reads being bucketed. Like @tschottdorf mentioned above, each map element would act a lot like a RWMutex, with the added benefit of deterministic ordering.

bdarnell commented:
> You would presumably queue all the writes back-to-back, i.e. there's a single list of batches. The downside here is that one write has to wait for a completely unrelated write to create its Batch, i.e. while a large DeleteRange is evaluated, a lot of writes to a different part of the Range has to just sit there.

That's the same as today, right (with all that work happening in raft command application)? So it's not a regression from today, just a case where we're not getting as much parallelism as we could. 

---

Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10084#-:-KUm5CoDFt9WFVdfkKz6:b-hgp9fn)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Yeah, you're right, that amounts to the same but we're presumably doing at least a little more work by piling multiple batches upon each other.

---

Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10084#-:-KUm8JRg1AQtkYwpdd-L:b3ulzh5)*
<!-- Sent from Reviewable.io -->

tbg committed:
[DNM] add actual keys read to command queue

The code is a mess and not meant to go in; don't bother leaving style comments.

This is an exploration for #6290. Briefly put, evaluation of commands before
proposing their result to Raft exercises the command queue in ways the current
code did not (which executes everything serially because it's on the Raft
goroutine).

In particular, the command queue now has to live up to its role of allowing
only commands which can be parallelized to be evaluated (and subsequently
proposed and applied).

What the command queue tracked so far are simply the main key ranges of the
requests in the batch, but that is not nearly enough.

To err on the side of being conservative, this PR presents an (unmergeably
hacky) attempt to classify what keys really need to be protected, as measured
by our testing coverage. As such, this list is not complete and in particular
it does not take into account reads of Replica state where the in-memory cache
is hit (though that could be caught by not accessing state directly but having
it go to disk every time; not done yet).

Major takeaways are:

- Splits need to lock the key range of the RHS descriptor for writes (we can
  get a little tricky with reads), but also touches all of the abort cache for
  reading. It also accesses the Replica state for the RHS, though that's
  a lesser concern.
- Every command needs to read FrozenStatus, RangeLastGCKey and writes
  MVCCStats. Unless we are smart about it, that already completely kills off
  any proposal parallelism. We will have to be smart about it because that's
  pretty bad. Especially the stats update is interesting because that is both
  frequent and a write operation. The obvious solution here is to make use of
  the commutativity of stats delta updates (though `ContainsEstimates` needs to
  be taken into account conservatively).
- Every transaction reads its AbortCacheKey and reads or writes its transaction
  record (see discussion in #6290 on why the latter makes sense).

The proposal's MaxLeaseIndex already makes sure that when we submit command
A to Raft before command B, then they can only apply in that order (mod
refurbishments, which can reorder and cause problems to this arguments, though
I think they can be overcome). So we could imagine a world in which the command
queue wouldn't commonly block commands on top of each other:

- a command arrives.
- checks the command queue; finds that it must wait for some prior commands to
  apply.
- instead of waiting, it evaluates these commands in the correct order and
  applies itself on top of that, entering itself in the command queue but
  immediately proposing to Raft a command whose successful application is
  contingent upon the previous commands applying as planned.
- once the command applies, leave the command queue (as usual)
- if the command does not apply (i.e. it or its dependencies got reordered),
  leave the command queue and enter it anew (instead of the previous internal
  refurbishment cycle).

The above is fairly half baked and challenging to implement, but so is tracking
every command's implicit writes; the above idea would in principle remove any
latency that isn't Raft or doing actual work, which is a great benefit.

tschottdorf commented:
@nvanbenschoten I'll send a PR splitting the command queue into two and base my work for cleaning this PR up on top of that, at which point I'd imagine addressing the most egregious performance penalties with your help.

bdarnell commented:
It seems to me that there is a large amount of work here and it would be significantly simpler (and not significantly worse in performance) to use a list of batches for the first version of proposer-evaluated KV. You're closer to this code than I am, though.

tschottdorf commented:
> It seems to me that there is a large amount of work here and it would be significantly simpler (and not significantly worse in performance) to use a list of batches for the first version of proposer-evaluated KV. You're closer to this code than I am, though.

I was on the fence about that too, but now it looks that while the cleanup direction might be a bit more work, the batch list road feels like taking a shortcut which kicks some of the work done the road. There are some unknowns (how to stack those batches, what's the performance of that really) and it's less amenable to straightforward optimization once in place since it marginalizes the command queue and lets us get away with the deficiencies discovered in this PR. Now that I've mapped them out, shouldn't be too hard to do some cleanup & end up where we want to be. That's my thinking, at least. Let's see how I feel tomorrow now that I'm tackling this.

tschottdorf commented:
This PR is obsolete given @bdarnell's work on proposer-evaluated KV.
